File: server/storage.py
import json
from pathlib import Path
from server.config import ASSETS_DIR


def save_asset(key: str, data: bytes) -> str:
    path = ASSETS_DIR / key
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(data)
    return f"/assets/{key}"


def save_json(key: str, data: dict) -> None:
    path = ASSETS_DIR / key
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(data, indent=2))


def load_json(key: str) -> dict | None:
    path = ASSETS_DIR / key
    if path.exists():
        return json.loads(path.read_text())
    return None

File: src/server/storage.ts
import { mkdir, writeFile, readFile } from 'fs/promises'
import path from 'path'
import { config } from './config.js'

const base = path.resolve(config.ASSETS_DIR)

export async function saveAsset(key: string, data: Buffer | string): Promise<string> {
  const fullPath = path.join(base, key)
  await mkdir(path.dirname(fullPath), { recursive: true })
  await writeFile(fullPath, data)
  // Return a URL path the Express static server will serve
  return `/assets/${key}`
}

export async function loadJson<T>(key: string): Promise<T | null> {
  try {
    const fullPath = path.join(base, key)
    const raw = await readFile(fullPath, 'utf-8')
    return JSON.parse(raw) as T
  } catch {
    return null
  }
}

export async function saveJson(key: string, data: unknown): Promise<void> {
  const fullPath = path.join(base, key)
  await mkdir(path.dirname(fullPath), { recursive: true })
  await writeFile(fullPath, JSON.stringify(data, null, 2))
}
